ExpressLRS - fixed slow initial connection

phobos- opened the pull request:
Fixes #11153 

Timer, IRQ and RX are initialized at different times during startup, causing the elrs timer to measure huge initial offset that causes the link to connect after approximately 30 seconds. In this PR we also stop the timer just before eeprom write and restart it just after to ensure proper offset measurements.

[betaflight_4.3.0_CRAZYBEEF4SX1280_7046c54fd.zip](https://github.com/betaflight/betaflight/files/7821257/betaflight_4.3.0_CRAZYBEEF4SX1280_7046c54fd.zip)
[betaflight_4.3.0_BETAFPVF4SX1280_7046c54fd.zip](https://github.com/betaflight/betaflight/files/7821258/betaflight_4.3.0_BETAFPVF4SX1280_7046c54fd.zip)

PSi86 commented:
betaflight_4.3.0_CRAZYBEEF4SX1280_a63af2aa3 does not fix the long sync time after power up.
I am using 250Hz with RX set to this specific rate as well: set expresslrs_rate_index = 1

sendorm commented:
Tried betaflight_4.3.0_CRAZYBEEF4SX1280_a63af2aa3. It doesn't fix the issue unless I set the cpu overclock to 120MHz. Same behavior as RC1.

phobos- commented:
@PSi86 @sendorm try this:
```
set dshot_bitbang = AUTO
save
```

PSi86 commented:
> @PSi86 @sendorm try this:
> 
> ```
> set dshot_bitbang = AUTO
> save
> ```

Tested; After changing the option as indicated by you, after power on the LQ was changing around between 0-3 (before: 0 until connected). Unplugged and replugged 7 times - same behaviour. Then after waiting aprox 30seconds the connection was established (lq:100) - just like before changing the option. But after this one complete sync it now instantly syncs directly after power up (tried around 20 times).

phobos- commented:
@PSi86 
> Then after waiting aprox 30seconds the connection was established (lq:100) - just like before changing the option. But after this one complete sync it now instantly syncs directly after power up (tried around 20 times).

First sync can take longer than usual when ExpressLRS settings are different from current (or default) settings stored in the FC. It can be circumvented by setting below cli variables to appropriate values for your setup:

```
expresslrs_rate_index = 0
Allowed range: 0 - 3

expresslrs_switch_mode = HYBRID
Allowed values: HYBRID, WIDE

expresslrs_model_id = 255
Allowed range: 0 - 255
```

Once set (or initially connected) the sync behaviour should be as you described.

PSi86 commented:
@phobos- As described in my ealier post, I already use the options you mention. The whoop was used on two flying sessions with around 20 batteries flown and it always took the 30 seconds to connect. The only thing which was then changed was to set `dshot_bitbang = AUTO` and then the behaviour described above occured. So the only way this makes sense is if the settings were not saved in eeprom of the quad sucessfully on all the earlier flights with dshot_bitbang = OFF. If this is expected then everything is fine - else there still might be something unnoticed going on in the code.

phobos- commented:
`diff all` only catches differences. You probably have the default value of 255 set.

P-I-Engineer commented:
> Tried betaflight_4.3.0_CRAZYBEEF4SX1280_a63af2aa3. It doesn't fix the issue unless I set the cpu overclock to 120MHz. Same behavior as RC1.

is your pid loop at 2k or 4k? at 2k i'm connecting very fast, at 4k, it's spotty.

PSi86 commented:
> `diff all` only catches differences. You probably have the default value of 255 set.

Ok, perfect if that is the default then there shouldn't be anything to save for the FC on the first sync. All values were already set in cli or left at default. In this case there is no real explanation for the observed behavior. But as it works now and nobody else made any similar observations its fine.

AndreaCCIE commented:
tested on two CRAZYBEEF4SX1280, it works to me

daleckystepan commented on src/main/rx/rx_spi.c:
```diff
@@ -67,6 +67,12 @@ static protocolDataReceivedFnPtr protocolDataReceived;
+static void nullProtocolStop(void) {}
+static protocolStopFnPtr protocolStop = nullProtocolStop;
+#endif
+
```
It is a little bit tricky to add `protocolStopFnPtr` this way. How about adding it as another property like `protocolSetRcDataFromPayloadFnPtr` and all others? You can initialize it to NULL or nullProtocolStop in the begging of `rxSpiSetProtocol`. Maybe it will add few bytes but it is more consistent solution.

phobos- replied:
Thanks for looking, I will fix it in a few hours

Zuldan commented:
Just a fyi, before Steves mashup PR was merged, my F411 with a Frsky D / SPI RX (4k pid / dshot300) took 1 second to establish a connection (however, I did see “BADRX/RXLOSS for half a second on the OSD). After Steves mashup was merged my Frsky D took 30 seconds to establish a connection (BADRX/RXLOSS on the whole time). Overclocking to 120Mhz resolved the issue completely with no “BADRX/RXLOSS” messages at all. Maybe there are some underlying issues and some of the newer code has aggravated the problem further.

AlessandroAU commented:
All seems good on my CRAZYBEEF4SX1280 boards

phobos- commented:
@Zuldan it may be related, although changes from this PR seem to fix the issue with ExpressLRS without the need for 120MHz overclock. Bear in mind in 4.3 (and older versions) overclock is default for F411 (108MHz).

PSi86 commented:
Please rebuild for testing with latest merges. Thank you!

haslinghuis commented:
Rebased on master:

[betaflight_4.3.0_STM32H743_8389a196e.zip](https://github.com/betaflight/betaflight/files/7835416/betaflight_4.3.0_STM32H743_8389a196e.zip)
[betaflight_4.3.0_STM32F7X2_8389a196e.zip](https://github.com/betaflight/betaflight/files/7835410/betaflight_4.3.0_STM32F7X2_8389a196e.zip)
[betaflight_4.3.0_STM32F405_8389a196e.zip](https://github.com/betaflight/betaflight/files/7835411/betaflight_4.3.0_STM32F405_8389a196e.zip)
[betaflight_4.3.0_STM32F411_8389a196e.zip](https://github.com/betaflight/betaflight/files/7835413/betaflight_4.3.0_STM32F411_8389a196e.zip)
[betaflight_4.3.0_STM32F745_8389a196e.zip](https://github.com/betaflight/betaflight/files/7835414/betaflight_4.3.0_STM32F745_8389a196e.zip)
[betaflight_4.3.0_STM32G47X_8389a196e.zip](https://github.com/betaflight/betaflight/files/7835415/betaflight_4.3.0_STM32G47X_8389a196e.zip)

hydra commented on src/main/rx/expresslrs_impl.h:
```diff
@@ -73,6 +73,7 @@ typedef struct elrsReceiver_s {
     bool alreadyFHSS;
     bool alreadyTLMresp;
     bool lockRFmode;
+    bool started;
```
Now that there are 4 boolean flags to record state, it feels like bit flags should be used instead.  i.e.  `uint8_t flags`  and 
```
#define ELRS_RX_STATE_FHSS (1 << 0)
#define ELRS_RX_STATE_TLM_RESPONDED (1 << 1)
#define ELRS_RX_STATE_LOCK_RF (1 << 2)
#define ELRS_RX_STATE_STARTED (1 << 3)
```

Thoughts?

phobos- replied:
Good idea, I'll refactor it later

hydra commented on src/main/rx/expresslrs.c:
```diff
@@ -408,13 +416,13 @@ static bool handleSendTelemetryResponse(void)
     if (receiver.connectionState != ELRS_DISCONNECTED && expressLrsEPRHaveBothEvents()) {
-        pl.rawOffsetUs = expressLrsEPRGetResult();
+        int32_t maxOffset = receiver.modParams->interval / 4;
+        pl.rawOffsetUs = constrain(expressLrsEPRGetResult(), -maxOffset, maxOffset);
```
This is the main fix of this PR right?

phobos- replied:
correct, together with L1066 changes

ctzsnooze commented:
@phobos - RC2 is out, and we may only have one more chance to fix any outstanding issues with ELRS SPI before final 4.3 release.  Please triple check that ELRS SPI is working properly on all supported boards, and let's see if we can finalise this PR in the next 7 days, if possible.

haslinghuis commented:
Rebased on master:

[betaflight_4.3.0_STM32G47X_2b6e27b34.zip](https://github.com/betaflight/betaflight/files/7836562/betaflight_4.3.0_STM32G47X_2b6e27b34.zip)
[betaflight_4.3.0_STM32H743_2b6e27b34.zip](https://github.com/betaflight/betaflight/files/7836563/betaflight_4.3.0_STM32H743_2b6e27b34.zip)
[betaflight_4.3.0_STM32F411_2b6e27b34.zip](https://github.com/betaflight/betaflight/files/7836559/betaflight_4.3.0_STM32F411_2b6e27b34.zip)
[betaflight_4.3.0_STM32F745_2b6e27b34.zip](https://github.com/betaflight/betaflight/files/7836561/betaflight_4.3.0_STM32F745_2b6e27b34.zip)
[betaflight_4.3.0_STM32F7X2_2b6e27b34.zip](https://github.com/betaflight/betaflight/files/7836557/betaflight_4.3.0_STM32F7X2_2b6e27b34.zip)
[betaflight_4.3.0_STM32F405_2b6e27b34.zip](https://github.com/betaflight/betaflight/files/7836558/betaflight_4.3.0_STM32F405_2b6e27b34.zip)

AlessandroAU commented:
[betaflight_4.3.0_CRAZYBEEF4SX1280_2d5f0fd42.zip](https://github.com/betaflight/betaflight/files/7836642/betaflight_4.3.0_CRAZYBEEF4SX1280_2d5f0fd42.zip)

JyeSmith commented:
Tested with rebased hex. No problems.

haslinghuis commented:
Firmware rebased on master:


[betaflight_4.3.0_STM32F7X2_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842587/betaflight_4.3.0_STM32F7X2_b22c825ef.zip)
[betaflight_4.3.0_STM32F405_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842588/betaflight_4.3.0_STM32F405_b22c825ef.zip)
[betaflight_4.3.0_STM32F411_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842589/betaflight_4.3.0_STM32F411_b22c825ef.zip)
[betaflight_4.3.0_STM32F745_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842590/betaflight_4.3.0_STM32F745_b22c825ef.zip)
[betaflight_4.3.0_STM32G47X_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842591/betaflight_4.3.0_STM32G47X_b22c825ef.zip)
[betaflight_4.3.0_STM32H743_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842592/betaflight_4.3.0_STM32H743_b22c825ef.zip)

JyeSmith commented:
> Firmware rebased on master:
> 
> [betaflight_4.3.0_STM32F7X2_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842587/betaflight_4.3.0_STM32F7X2_b22c825ef.zip) [betaflight_4.3.0_STM32F405_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842588/betaflight_4.3.0_STM32F405_b22c825ef.zip) [betaflight_4.3.0_STM32F411_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842589/betaflight_4.3.0_STM32F411_b22c825ef.zip) [betaflight_4.3.0_STM32F745_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842590/betaflight_4.3.0_STM32F745_b22c825ef.zip) [betaflight_4.3.0_STM32G47X_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842591/betaflight_4.3.0_STM32G47X_b22c825ef.zip) [betaflight_4.3.0_STM32H743_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842592/betaflight_4.3.0_STM32H743_b22c825ef.zip)

Can you please include the required crazybee sx1280 legacy target.

haslinghuis commented:
[betaflight_4.3.0_CRAZYBEEF4SX1280_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842747/betaflight_4.3.0_CRAZYBEEF4SX1280_b22c825ef.zip)

phobos- commented:
@haslinghuis thanks for posting the targets. I just rebased against master and fixed review comments. I will address @hydra 's suggestion in a separate PR - it is a refactoring change and can be included in 4.4. This one should be ready.

PSi86 commented:
I hate to say it because I'd love to see this working but current master build [2616](https://ci.betaflight.tech/job/Betaflight/2616/) works way better for me than CRAZYBEEF4SX1280_b22c825ef, which is linked here. Double checked the dshot settings - bitbang is used both times. Still using 250Hz link rate and no telemetry.

MLutzAT commented:
> [betaflight_4.3.0_BETAFPVF4SX1280_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842885/betaflight_4.3.0_BETAFPVF4SX1280_b22c825ef.zip) [betaflight_4.3.0_CRAZYBEEF4SX1280_b22c825ef.zip](https://github.com/betaflight/betaflight/files/7842747/betaflight_4.3.0_CRAZYBEEF4SX1280_b22c825ef.zip)

Can confirm that both of these images work fine for me! One on my Meteor 65 ELRS and the other on my Mobula 6 ELRS. Bitbang on (is it still necessary, as this was also suggested me from Steve solving #11246 ,before a regular fix was done?), bidirectional DSHOT300, RPM filtering working with BlueJay on the ESCs, 4K PID Loop. TX is BetaFPV 2,4 Module (flashed to ELRS 2.0.1) mounted on my Taranis running latest OpenTX. Connection is established within a few seconds!

haslinghuis commented:
@MLutzAT can you please also test with latest dev build available in configurator?

PSi86 commented:
> @JyeSmith @MLutzAT can you please also test with latest dev build available in configurator?
> 
> @PSi86 It seems the issue being addressed here has already been fixed in master now: [#11271 (comment)](https://github.com/betaflight/betaflight/issues/11271#issuecomment-1010454330)

Your link is pointing to the OSD profile thread, however my testing was related to the sync speed on power up. But yes, I have the feeling too that the originating issue was solved in an already merged PR. Would be interessting to test this PR built against master once more.

haslinghuis commented:
Rebased on master:

[betaflight_4.3.0_BETAFPVF4SX1280_b4d74601c.zip](https://github.com/betaflight/betaflight/files/7850617/betaflight_4.3.0_BETAFPVF4SX1280_b4d74601c.zip)
[betaflight_4.3.0_CRAZYBEEF4SX1280_b4d74601c.zip](https://github.com/betaflight/betaflight/files/7850618/betaflight_4.3.0_CRAZYBEEF4SX1280_b4d74601c.zip)

PSi86 commented:
Nice! With CRAZYBEEF4SX1280_b4d74601c I get instant sync and stable connection at 250Hz. Flew only 3 batteries now but plugged in around 15 times. - More testing tomorrow.

CMartin249 commented:
> Can confirm that both of these images work fine for me! One on my Meteor 65 ELRS and the other on my Mobula 6 ELRS. Bitbang on (is it still necessary, as this was also suggested me from Steve solving #11246 ,before a regular fix was done?), bidirectional DSHOT300, RPM filtering working with BlueJay on the ESCs, 4K PID Loop. TX is BetaFPV 2,4 Module (flashed to ELRS 2.0.1) mounted on my Taranis running latest OpenTX. Connection is established within a few seconds!

Curious too what the recommended dshot_bitbang setting is for CRAZYBEEF4SX1280 on Mobula6 ELRS. DMA SHOW seems to be the same with it set to ON or AUTO. Unless it is OFF, the Gyro can't use DMA. Seems to fly fine with it either ON or OFF. To be honest, not sure I'd be able to tell. 

Tested both betaflight_4.3.0_CRAZYBEEF4SX1280_b22c825ef and betaflight_4.3.0_CRAZYBEEF4SX1280_b4d74601c firmware versions briefly today. b22c825ef does not connect to tx when pid loop is 4k, even after changing cpu_overclock to 120MHZ. b4d74601c does sync/connects fine with 4k pid loop.

phobos- commented:
@PSi86 @MLutzAT I am not surprised Steves fix improves things. I developed ExpressLRS code on bf version from before scheduler changes and it worked perfectly there. I still think changes in this PR improve the robustness of the connection further. That should be especially visible when doing EEPROM write - current master will take slightly longer to connect after EEPROM write compared to this PR.

mard89 commented:
I'm having issues with random fail safes on 500hz. While b22c825ef solved issues with OSD, this issue seems to persist. I'm running 4K and DSHOT300 on the betafpv board which is overclocked to 120mhz.

The fail safes don't occur every time. I think once every +/-4 to 6 packs. But once I get one one, chances are I get another one on the next pack. RSSI keeps updating before and during the fail safe and never drops below -55dbm.

PSi86 commented:
Did more test-flights with CRAZYBEEF4SX1280_b4d74601c and it works perfectly.
Link is established extremely quick. Link is also very stable. No OSD issues observed. Nothing negative to report.

PSi86 commented:
@mard89 To my knowledge 500Hz was never recommended or found to be stable with the SPI RX. Please correct me if I'm wrong.

PSi86 commented:
@ctzsnooze Thanks for the correction. 500Hz always gave me failsafes, except with the realtime priority patch. Going to test again tomorrow with the current FW.

phobos- commented:
@mard89 
> But once I get one one, chances are I get another one on the next pack
This code cannot impact subsequent power cycles, it must be something that persists - that suggests a Handset issue. 

If you have an external ELRS receiver (EP2 would be ideal) I suggest soldering it in and repeating the test on the same settings and the same quad.

I just ran the link at 500hz for 1 hour without a failsafe on the bench. Please try the same. To verify if it failsafed use `set debug_mode = RX_EXPRESSLRS_SPI` and note the `debug 0` value in the beginning of the test and in the end of the test. Usually it shows 0-2 in the beginning and the value should not change by the end of the test. You can also verify the value incrementing by forcing a failsafe by disconnecting and connecting the TX or doing an EEPROM write.

PSi86 commented:
500Hz is working just like @ctzsnooze said. Happy to to be wrong. 10 Flights - no issue. Currently not using bidir dshot however.
@ctzsnooze Do you use bidir dshot or is further testing required for that option?

mard89 commented:
I just flew 4 packs on b4d74601c and had 3 failsafes. I was under the impression that this time the failsafes were preceded by smaller hiccups instead of suddenly dropping out of the sky. And also the time between the failsafe en reestablishment of the link appeared to be (even) shorter.

@PSi86 500hz was working on ELRS v1 on whatever version betaflight betafpv shipped the board with. I started running into this problem after I upgraded to 4.3 rc1.

@phobos- I wasn't suggesting that the problem was persistent. My remark was simply to explain that there is no clear pattern to the occurrence of the failsafes; they occur randomly.
The combination of the RM tx16s and HM module was working perfectly in combination with other quads that are using EP1 and EP2 receivers. Also, the dbm value is still updating when I failsafe (even though the LQ drops to 0) and never gets bellow -55dbm.
I have not been flying any of the EP1 or EP2 quads for about 3 weeks now due to bad weather, so I will try to induce a failsafe on the bench while running set debug_mode = RX_EXPRESSLRS_SPI and try the same for different quads to make certain that my handset or module isn't the culprit.

mard89 commented:
Just done with checking my handset and the betafpvsx1280 whoop on the bench. Watched the link of a EP1 receiver using the `CRSF_LINK_STATISTICS_UPLINK` debug mode for over an hour. Nothing but stable lines.
Next I connected the betafpv board using `RX_EXPRESSLRS_SPI'. After about 10 minutes I got the first fail safe, and another one at around 15 minutes.

![Screenshot_20220113_190828](https://user-images.githubusercontent.com/51527720/149393507-efaaf912-681c-4cb8-bcc9-38520d7ad3db.png)

PSi86 commented:
Tested 500Hz with bidir dshot: CPU usage is around 15% higher (with rpm filtering), so around 55-60%. No failsafes here.
@mard89 What is your cpu utilisation? Do you have any additional peripherals configured? Sadly I dont have the BetaFPV Board but the Crazybee, so I cannot match your setup exactly. Did you also try without the overclocking? Someone mentioned here in another PR that the overclocking made his FC behave erratically.

SantiCO19 commented:
Been testing HM2g4 for the last 90min. Firmware provided by the latest BFC nightly is not flyable. Failsafe instantly using the similar config as everyone here, bidir/8/4/dsh300/500hz. OC to 120MHz doesn't help. I then flashed b4d74601c, same basic config, plus OC 120MHz, and Blackbox logging on serial uart 1. Debugging mode set at standard Gyro_Scaled because there's not an ELRS debug option available in the drop down. Okay, so link establishes quickly.  Armed and lifted off. I completed many short flights, playing with  PIDs in between. Also landing to adjust acc cal via stick inputs. The write still breaks the link, but wow good job fixing the code so that it at least reestablishes almost instantly. (I've struggled with this so much until now, that i find the current behavior to be such an improvement, it feels more than acceptable. I can once again adjust my acc cal relatively quickly.) Okay, so after around third pack I decided to play with presets. I loaded some 3" tune, and I loaded ELRS 500hz. Had to then adjust back to spi elrs, so i don't know if that preset helped me or not. So anyway, right after loading the presets, first attempt to fly, right away I got my first quick failsafe before even arming. Reconnected within5-6 seconds. I assumed maybe cpu use was causing it, as I'm sitting around 50-60 i think, so i let it perhaps stabilize about 10-15 seconds. It FS one more time in the first 5 secs, quickly recovering, and then seems stable after the another 10-12s or so. I armed and lifted off, playing much the same as previously, short flights, adjust acc cal, this is all indoor btw. So over the course of another pack or two, I FS a number of times. Sometimes took up to 15+seconds I'd guess, to come back. I have some funny looking logs, where all traces abruptly stop, each punctuated by a little square, followed by short blank section with no traces, then all traces reappear, with little square indicating where they came back online. Unfortunately i can't correlate this peculiarity with the failsafes I had, but seems like it could match up. I then have logs with disarm:failsafe indicated, and I thought this might be during the eeprom write after stick command, but the stick positions don't match up with that hypotheses. I made so many land/takeoff/land/adjust/power off/plug usb/power on/land/take off etc., that I found i didn't keep track of the logs well. Sorry. I still forget i have it sometimes. Anyway, i really wanna go try a real fpv flight, so i think i might risk it and try. Perhaps ill reflash and avoid the preset this time, as im pretty sure i  had no FS during low impact, indoor test flights, with the firmware from this thread, until I used a preset. I will try to get some more useful answers on this matter over the next few packs. If someone would like to outline some specifics that theyd like to see logged, id be happy to try it. FWIW, this is 3" ultralight toothpick build on the HM2g4 5in1 of course. Also, it seems strange the ELRS debugs are missing but there is frsky_spi, ghost, etc., but i dont really know the criteria for those options to appear there, so maybe ees-normal. Ill check in back here regularly, as i know time seems to be of the essence.

SantiCO19 commented:
> Tested 500Hz with bidir dshot: CPU usage is around 15% higher (with rpm filtering), so around 55-60%. No failsafes here. @mard89 What is your cpu utilisation? Do you have any additional peripherals configured? Sadly I dont have the BetaFPV Board but the Crazybee, so I cannot match your setup exactly. Did you also try without the overclocking? Someone mentioned here in another PR that the overclocking made his FC behave erratically.

I have the crazybee too. I'm running openlager in addition to the similar settings as yours. No apparent issues through many short flights over one pack. I started failsafing **only after application of two presets**. Didn't happen often though, but seemed to happen consistently there after.

PSi86 commented:
@SantiCO19 could you try 4k/4k and without the logging? Also check the `dma show` in cli to see, if dshot bitbang is shown.

mard89 commented:
Out of academic interest I tried 7c9e5ac8e from #11154
It took a whole lot longer to get a failsafe on the bench, and I only got 1 in over an hour of testing. It did take longer for the link to re-establish, and the failsafe counter also wasn't triggered (note the 0 at debug 0)

![Screenshot_20220114_130143](https://user-images.githubusercontent.com/51527720/149522291-cfb76bd8-0114-466a-865d-7c050bb23f64.png)

I also tried flying this build. Not my smartest idea ever, because the failsafe wasn't triggered when the link dropped and it took about a second or 5 before me flicking the disarm switch was processed.



> @SantiCO19 could you try 4k/4k and without the logging? Also check the `dma show` in cli to see, if dshot bitbang is shown.

The gyro update frequency is locked to the native update frequency since betaflight 4.2. Which in most cases is 8K.

> I started failsafing **only after application of two presets**. Didn't happen often though, but seemed to happen consistently there after.

I also applied the the ELRS 500hz link preset. But I do not think that this is linked to the failsafes. I was having failsafes before I applied the preset.

SantiCO19 commented:
> @SantiCO19 could you try 4k/4k and without the logging? Also check the `dma show` in cli to see, if dshot bitbang is shown.

So, yesterday I failsafed into some snow about 20 minutes after my last comment. Camera fried instantly. Haven't messed with the board. But I have another HM2g4, so I set it up, and have not got any FS. Not sure how to change 8k to 4k for gyro. dma show output looks like this: `Currently active DMA:
--------------------
DMA1 Stream 0: SPI_MISO 3
DMA1 Stream 1: FREE
DMA1 Stream 2: FREE
DMA1 Stream 3: SPI_MISO 2
DMA1 Stream 4: SPI_MOSI 2
DMA1 Stream 5: SPI_MOSI 3
DMA1 Stream 6: FREE
DMA1 Stream 7: FREE
DMA2 Stream 0: ADC 1
DMA2 Stream 1: DSHOT_BITBANG 2
DMA2 Stream 2: FREE
DMA2 Stream 3: FREE
DMA2 Stream 4: FREE
DMA2 Stream 5: FREE
DMA2 Stream 6: FREE
DMA2 Stream 7: FREE`

SantiCO19 commented:
> I also applied the the ELRS 500hz link preset. But I do not think that this is linked to the failsafes. I was having failsafes before I applied the preset.

I also applied the 533 Trainer Karate preset. I did not have FS until I did that, but maybe I didn't give enough time. I removed the the preset, and tried a real flight. About 50m out, the quad FS face first into snow, and fried the camera instantly. I recovered it and unplugged the battery, and haven't really wanted to look at the board yet.  A second board seems to be working, but my testing has been limited. I'm trying to do some more now.

SantiCO19 commented:
Okay, I'm just flying around inside. 5-6 mins in and I got my first FS. No presets, etc..

PSi86 commented:
@SantiCO19 @mard89 Do the Failsafes also occur at 250Hz link speed?

SantiCO19 commented:
> @SantiCO19 @mard89 Do the Failsafes also occur at 250Hz link speed?

ill check

wang322588 commented:
I tested betafpv board.At 500Hz on OSD showed "rssi dbm" not "fail safe" and reconntected soon. accured every 2-3 min. I think maybe CPU load is 60% then I turned off Bidir Dshot .its 38% and seems the disconnection never happened.

ctzsnooze commented:
@wang322588 please repeat your 500Hz link test with `set cpu_overclock = 120`, if not already overclocked

haslinghuis commented:
Final testing (rebased on master):

[betaflight_4.3.0_STM32F7X2_90cd5fb0c.zip](https://github.com/betaflight/betaflight/files/7876262/betaflight_4.3.0_STM32F7X2_90cd5fb0c.zip)
[betaflight_4.3.0_STM32F405_90cd5fb0c.zip](https://github.com/betaflight/betaflight/files/7876263/betaflight_4.3.0_STM32F405_90cd5fb0c.zip)
[betaflight_4.3.0_STM32F411_90cd5fb0c.zip](https://github.com/betaflight/betaflight/files/7876264/betaflight_4.3.0_STM32F411_90cd5fb0c.zip)
[betaflight_4.3.0_STM32F745_90cd5fb0c.zip](https://github.com/betaflight/betaflight/files/7876266/betaflight_4.3.0_STM32F745_90cd5fb0c.zip)
[betaflight_4.3.0_STM32G47X_90cd5fb0c.zip](https://github.com/betaflight/betaflight/files/7876267/betaflight_4.3.0_STM32G47X_90cd5fb0c.zip)
[betaflight_4.3.0_STM32H743_90cd5fb0c.zip](https://github.com/betaflight/betaflight/files/7876268/betaflight_4.3.0_STM32H743_90cd5fb0c.zip)

JyeSmith commented:
Crazybee confirmed with a Hoover test on each RF rates.

CMartin249 commented:
Nice work! Flew several packs on b4d74601c and 90cd5fb0c indoors with no fail safes. I do get random/brief telemetry loss/recovered warnings on TX. Going into CLI always generates a quick telemetry loss/recovered warning. HM Mob6 ELRS SPI 500Hz/bidir/8k/4k/ds300/BitBangON/108MHZ-OC.

sendorm commented:
@CMartin249 Telemetry lost may also be due to elrs code itself. There is a pr for this in elrs. What baud rate are you using. Was it megabauds?

mard89 commented:
I'm still getting failsafes with 90cd5fb0c. The initial connection and reestablishment of the link after a failsafe is also taking longer than b4d74601c.

![Screenshot_20220116_101859](https://user-images.githubusercontent.com/51527720/149654286-421d0f5f-c296-4415-8430-9b5a7d9d51d9.png)



> @wang322588 please repeat your 500Hz link test with `set cpu_overclock = 120`, if not already overclocked

I can confirm that overclocking to 120mhz doesn't fix the problem; all my tests have been done at 120mhz.

AndreaCCIE commented:
@mard89 I didn’t test every single build, but the only issue I had was the long time to establish the link. No failsafe on 2 happymodel aio SPI in 2 weeks and 40ish packs. Are you sure it’s not a hardware problem? Have you tried a different board?

mard89 commented:
> @mard89 I didn’t test every single build, but the only issue I had was the long time to establish the link. No failsafe on 2 happymodel aio SPI in 2 weeks and 40ish packs. Are you sure it’s not a hardware problem? Have you tried a different board?

While I don't have a 2nd board for testing to properly rule out a hardware issue, there are some indication that it isn't. namely: 500hz was working on the ELRS v1 build the board shipped with; the failsafes were introduced when upgrading to 4.3rc1 (while upgrading to ELRS v2). And different builds behave differently: the occurrence/time between failsafes differs between builds, as does the time for the link to reestablish (independent of the initial connection time).

I do have an EP1 coming in to, hopefully, confirm that it is the SPI receiver.

CMartin249 commented:
> @CMartin249 Telemetry lost may also be due to elrs code itself. There is a pr for this in elrs. What baud rate are you using. Was it megabauds? 

Yes. EdgeTX 2.5 @3.7m max bauds(Jumper T16 Pro).

sendorm commented:
> 

Can you try with 400k? If my suspicion is right you won’t get telemetry lost anymore.

hydra commented:
I cherry-picked this into my H7RF `bf-h7rf-wip-3` target branch, here https://github.com/spracing/betaflight/tree/bf-h7rf-wip-3  See PR #11110

The H7RF seems to connect pretty quickly, but doesn't recover quickly after issuing `tasks` in the CLI which causes the link to drop too.

phobos- commented:
@mard89 can you open an issue and describe your failsafe problems?

SantiCO19 commented:
> @mard89 I didn’t test every single build, but the only issue I had was the long time to establish the link. No failsafe on 2 happymodel aio SPI in 2 weeks and 40ish packs. Are you sure it’s not a hardware problem? Have you tried a different board?

I have 5 HM 2G4. I share very similar issues to @mard89 on all of them. These boards were pretty awesome at release, and I'm sorry to say the overall experience has been downhill since. I am a 1S performance enthusiast that was averaging between 10 and 15 packs a day, and I'd easily fly 40+ packs on the weekend, but it's become more and more difficult to enjoy these builds. ELRS over SPI, when first released (pre V1.0 I believe,) was fantastic, and played a big role in my decision to go all in on ELRS. I dont regret a thing. I just cant fly these boards anymore, for now, and it's a shame. I invested in them based on the original experience, and they are something else now. For me, I think the solution will be to revert these boards back to older firmware, and commit a spare tx module them. F4 boards are clearly getting left behind, and I think that may be what we are actually dealing with here, perhaps a casualty of rapid advancement. I wonder in anything in the code was compromised, in order to meet deadlines, and hurry out a release. If that happened, I would like to think that top priority for the very next minor update would be to immediately go back, and optimize the things that were rushed, but really, I could just go back and stay on older firmware. I have no doubt there are still people flying 3.5.7. Actually, the HM 900 ELRS board was released with 3.5.7! I have a few of them. With F3 boards, I feel like I saw it coming. I hadn't bought an F3 board in 6 months, or a year, when they were retired. With F4 boards, I'm a little caught off guard, with stacks of brand new F4's. Anyway, feel free to delete this comment if needed.

haslinghuis commented:
He guys please also check https://github.com/betaflight/betaflight/pull/11330

hydra commented:
@SantiCO19 Personally I've not been recommending F4's to anyone, ever.  The F7 is like a fixed F4 with a cache and is a much better fit for our flight-controller needs, and the H7 is like an F7 on steroids.

I fully agree with your recent experience, I don't feel at all confident in the current BF `master` branch and have been seeing, and fixing, and trying to fix numerous issues as a result of the recent changes that happened before ELRS 2.x was merged.  The ELRS 2.x code itself is fine, and I have a 2.x branch for the H7RF which is perfect in terms of link stability, which doesn't include some of the recent gyro/scheduler changes which is where things started going wrong for me.

Working:
https://github.com/spracing/betaflight/commits/bf-h7rf-and-config-in-memory-mapped-flash-and-elrs-v2

Many issues:
https://github.com/spracing/betaflight/commits/bf-h7rf-wip-5

Feel free to check the base of those branches to find the commits that started causing issues.

AndreaCCIE commented:
@hydra, there are almost no F7 options in the tiny-whoop/toothpick market, especially with integrated ELRS.
We can't all fly 5 inches, unfortunately. I'm not part of the development team, but I understand the frustration on both sides.
I had no issues testing the current master, but I have a limited (6) amount of FCs.
If it's not ready, it's not ready.

sendorm commented:
Well the thing gets more complicated when you put the tx firmware into consideration too. With f411 elrs oc’ed to 120MHz with 4.3rc2 with edgetx 2.6 and with baud rate 400k. I get no errors. While rx losses at 3.75MHz with t-lite. Any thing different in these options might/will create another outcome.
